canton/community/ledger/ledger-api-core/src/main/scala/com/digitalasset/canton/platform/apiserver/services/admin/ApiPackageManagementService.scala

```diff
@@ -131,57 +133,138 @@ private[apiserver] final class ApiPackageManagementService private (
     } yield dar
   }
 
-  private def validateUpgrade(upgradingPackage: Dar[(Ref.PackageId, Ast.Package)])(implicit
-      loggingContext: LoggingContextWithTrace
+  private def lookupDar(pkgId: Ref.PackageId)(implicit
+                                                  loggingContext: LoggingContextWithTrace
+  ): Future[Option[(Ref.PackageId, Ast.Package)]] = {
+    for {
+      optArchive
+        <- packagesIndex.getLfArchive(pkgId)
+      optPackage
+        <- Future.fromTry {
+        optArchive.traverse(Decode.decodeArchive(_))
+          .handleError(Validation.handleLfArchiveError)
+      }
+    } yield optPackage
+  }
+
+  private def existingVersionedPackageId(dar: (Ref.PackageId, Ast.Package)): Option[Ref.PackageId] = {
+    val (_, pkg) = dar
```

Comment:
In this and following two methods, you pass the `dar: (Ref.PackageId, Ast.Package)` but only actually need the Ast.Package. Can you simplify then?